ci/lint-commit-msg.py

```diff
@@ -4,116 +4,143 @@
     sys.stdout.flush()
-    ret = subprocess.run(cmd, check=True, stdout=subprocess.PIPE, text=True)
+    ret = subprocess.run(cmd, check=True, stdout=subprocess.PIPE, encoding="utf-8", text=True)
     return ret.stdout
 
-lint_rules: dict[str, tuple[Callable, str]] = {}
+lint_rules: dict[str, LintRule] = {}
 
 # A lint rule should return True if everything is okay
-def lint_rule(description: str):
-    def f(func):
+def lint_rule(description: str) -> Callable[[LintFunction], None]:
+    def f(func: LintFunction) -> None:
         assert func.__name__ not in lint_rules
         lint_rules[func.__name__] = (func, description)
     return f
 
-def get_commit_range() -> Optional[str]:
-    if len(sys.argv) > 1:
-        return sys.argv[1]
-    # https://github.com/actions/runner/issues/342#issuecomment-590670059
-    event_name = os.environ["GITHUB_EVENT_NAME"]
-    with open(os.environ["GITHUB_EVENT_PATH"], "rb") as f:
-        event = json.load(f)
-    if event_name == "push":
-        if event["created"] or event["forced"]:
-            print("Skipping logic on branch creation or force-push")
-            return None
-        return event["before"] + "..." + event["after"]
-    elif event_name == "pull_request":
-        base = event["pull_request"]["base"]["sha"]
-        head = event["pull_request"]["head"]["sha"]
-        return f"{base}..{head}"
-    return None
-
-def do_lint(commit_range: str) -> bool:
-    commits = call(["git", "log", "--pretty=format:%H %s", commit_range]).splitlines()
+def get_pr_commits() -> list[tuple[str, list[str]]]:
+    pr_number = os.environ["GITHUB_REF"].split("/")[2]
+    api_url = os.environ["GITHUB_API_URL"]
+    repo = os.environ["GITHUB_REPOSITORY"]
+
+    url = f"{api_url}/repos/{repo}/pulls/{pr_number}/commits"
+    req = request.Request(url)
+    req.add_header("Accept", "application/vnd.github+json")
+    if "GITHUB_TOKEN" in os.environ:
+        req.add_header("Authorization", f"token {os.environ['GITHUB_TOKEN']}")
+
+    with request.urlopen(req) as response:
+        commits = json.load(response)
+
+    res = []
+    for commit in commits:
+        sha = commit["sha"]
+        message = commit["commit"]["message"]
+        res.append((sha, message.splitlines()))
+
+    return res
+
+def get_commits() -> list[tuple[str, list[str]]]:
+    if os.environ.get("GITHUB_EVENT_NAME") == "pull_request":
+        return get_pr_commits()
+
+    commit_range = sys.argv[1] if len(sys.argv) > 1 else None
+    if not commit_range:
+        return []
+
+    commits = call(["git", "log", "-z", "--pretty=format:%H%n%B", commit_range]).split("\x00")
+    res = []
+    for commit in commits:
+        if not commit.strip():
+            continue
+        sha, message = commit.split("\n", 1)
+        res.append((sha, message.splitlines()))
+
+    return res
+
+def lint_commit_message(body: list[str]) -> list[str]:
+    failed = []
+    assert len(body) > 0, "Commit message must not be empty"
```

Comment:
empty commits are possible in git so this shouldn't be an assert

Reply:
If someone hoes out of their way to create such commits, it's on them and can very well just assert instead of giving them nice output. git doesn't allow empty messages by default.